packages/block-editor/src/components/link-control/index.js

```diff
@@ -0,0 +1,219 @@
+							} ) }
+
+						>
+							{ suggestion.type.toLowerCase() === 'url' && (
+								<Icon className="block-editor-link-control__search-item-icon" icon="admin-site-alt3" />
+							) }
+							<span className="block-editor-link-control__search-item-header">
+								<span className="block-editor-link-control__search-item-title">
+									<TextHighlight text={ suggestion.title } highlight={ inputValue } />
+								</span>
+								<span className="block-editor-link-control__search-item-info">{ suggestion.info || filterURLForDisplay( safeDecodeURI( suggestion.url ) ) || '' }</span>
+							</span>
+							<span className="block-editor-link-control__search-item-type">{ suggestion.type.toLowerCase() || '' }</span>
+						</button>
+					) ) }
+				</div>
+			</div>
+		);
+		/* eslint-enable react/jsx-key */
+	};
+
+	const LinkControlAdditionalSettings = () => (
+		<div className="block-editor-link-control__settings">
+			<ToggleControl
+				label={ __( 'Open in New Tab' ) }
+				onChange={ partial( onSettingChange, 'new-tab' ) }
+				checked={ linkSettings[ 'new-tab' ] } />
+			{ isFunction( renderAdditionalSettings ) && renderAdditionalSettings() }
+		</div>
+	);
+
+	return (
+		<URLPopover
+			className="block-editor-link-control"
+			onClose={ closeLinkUI }
+		>
+			<div className="block-editor-link-control__popover-inner">
+				<div className="block-editor-link-control__search">
+
+					{ ! isEditingLink && (
+						<div
+							className={ classnames( 'block-editor-link-control__search-item', {
+								'is-current': true,
+							} ) }
+						>
+							<span className="block-editor-link-control__search-item-header">
+
+								<ExternalLink
+									className="block-editor-link-control__search-item-title"
+									href={ currentLink.url }
+								>
+									{ currentLink.title }
+								</ExternalLink>
+								<span className="block-editor-link-control__search-item-info">{ currentLink.info || filterURLForDisplay( safeDecodeURI( currentLink.url ) ) || '' }</span>
+							</span>
+							<span className="block-editor-link-control__search-item-type">{ currentLink.type.toLowerCase() || '' }</span>
+							<IconButton className="block-editor-link-control__search-item-action block-editor-link-control__search-item-action--edit" icon="edit" label={ __( 'Edit' ) } onClick={ onStartEditing } />
+						</div>
+					) }
+
+					{ isEditingLink && (
+						<form>
+							<URLInput
+								className="block-editor-link-control__search-input"
+								value={ inputValue }
+								onChange={ onInputChange }
+								autocompleteRef={ autocompleteRef }
+								onKeyDown={ ( event, suggestion ) => {
+									stopPropagationRelevantKeys( event );
+									if ( event.keyCode === ENTER ) {
+										onLinkSelect( event, suggestion );
+									}
+								} }
+								onKeyPress={ stopPropagation }
+								placeholder={ __( 'Search or type url' ) }
+								renderSuggestions={ renderSearchResults }
```

Comment:
My concern with this approach where custom UI can be built for the autocomplete suggestions is that it could lead to some inaccessible implementations out there in the wild (and possibly also internally).

`URLInput` should probably be completely in charge of how they render, but allow for some variation through props (e.g. inline/in a popover).

Reply:
> where custom UI can be built for the autocomplete suggestions is that it could lead to some inaccessible implementations

To some extend, I don't disagree with this and it's important that you raise this as a concern.

Clearly exposing an API which affords custom render implementations will always open up the possibility of poor coding practices.

That said, you will note we expose arguments to the render prop which correspond to the props that need to be added to the list and list item elements in the search suggestions list. By spreading these props onto the correct components in our custom render function we get the benefits of accessibility for free.

Also, I would highlight that even _without_ this render prop, inaccessible implementations are still _easily_ possible. For example, robbed of the possibility of implementing a custom UI for search suggestions, one might end up abusing the `additionalControls` prop of `URLPopover` to render out search suggestions which are not associated with the search box (currently there is a ARIA relationship between input and suggestions list). I've already seen [something similar done in the Image link edit render](https://github.com/WordPress/gutenberg/blob/0ee2b784c10ace28f6c84d079ec601ff80a76199/packages/block-library/src/image/edit.js#L214).

Personally, I'm leaning towards a render prop and providing maximum control when needed. With the props pattern discussed above and providing good solid documentation, we can head off some of the concerns you raise. As always, I'm open to being told I'm going against the grain here.

Reply:
It seems like a much more complicated pattern. What's the use-case beyond this?

Using separate components internally in the `URLInput` feels cleaner to me and has less concerns.

From a design point of view, there might be a desire to bring the existing autocomplete list more in line with this new design anyway.

Reply:
> It seems like a much more complicated pattern. 

It is relatively complicated, but I can't see an alternative that allows us to customise the rendering of search results. If you can please let me know 😄 

> Using separate components internally in the URLInput feels cleaner to me and has less concerns.

Breaking `URLInput` into separate components might be a nice refactor, but I don't follow how this will help us with customising the rendering of the search results to achieve the design required by the `LinkControl` component.

> ...there might be a desire to bring the existing autocomplete list more in line with this new design anyway.

For me (and without intending to sound facetious) the key term here is _"might"_. 

My instinct is that changing the internals and design of `URLInput` (or even `Autocomplete`) now to be more inline with the design @shaunandrews has created will block this `LinkControl` PR from landing and will require a wider discussion.

I'm happy to do this but [@mtias has already suggested](https://github.com/WordPress/gutenberg/issues/17557#issuecomment-544950967) we move towards a v1 of `LinkControl` under "experimental" flag as soon as possible. With my "expdient" hat on, it seems like a render prop is the only way to afford the customisation we require.

Again, I'm happy to be told I'm wrong, but as it stands I can't see a way to achieve the simple solution you're suggesting.

Reply:
> My instinct is that changing the internals and design of URLInput (or even Autocomplete) now to be more inline with the design @shaunandrews has created will block this LinkControl PR from landing and will require a wider discussion.

I think that's twisting my words. I didn't suggest the new design should be the default at all, I'm saying that the exposed API here is not needed and in my opinion shouldn't exist. I haven't seen it mentioned what the use case is for customised suggestions lists and it seems like this is being added without any of that reasoning. My feeling is that as a first step It'd be far more preferable to provide a simple api like:
```
<URLInput
  __experimentalInlineSuggestions
  // ...
/>
```

which causes the component to render in the new style. 

> For me (and without intending to sound facetious) the key term here is "might".

The title of the PR is 'Unified Link creation interface', which says it all, really. There's always been a strong desire to make this interface more consistent. I know this is considered an iteration on the path to consistency, but that doesn't mean we shouldn't be concerned about the public APIs it intends to introduce, or whether it's actually going to be easier to implement across the editor.

Reply:
I appreciate your candid feedback Daniel.

> I think that's twisting my words.

I'm sure you know that was not my intention. 

> I didn't suggest the new design should be the default at all

This is what I _misunderstood_ from your comments. Easily done.

> I'm saying that the exposed API here is not needed and in my opinion shouldn't exist.

I understand and acknowledge this. I think there is also an argument that it _does_ need to exist. 🤷‍♂ 

> I haven't seen it mentioned what the use case is for customised suggestions lists and it seems like this is being added without any of that reasoning.

Perhaps this is the issue. I thought we understood that this render prop is needed because by default `URLInput` outputs suggestions that look like this:

<img width="373" alt="Screen Shot 2019-10-23 at 12 17 57" src="https://user-images.githubusercontent.com/444434/67387636-3338ce00-f58f-11e9-8c9d-b815d74a6ae0.png">

...which is quite a long way from what the design requires:

<img width="755" alt="Screen Shot 2019-10-22 at 14 49 16" src="https://user-images.githubusercontent.com/444434/67387670-40ee5380-f58f-11e9-8f1e-a4669c339e2e.png">

Note for example:

* Additional meta info.
* Highlighting of search terms via `TextHighlight`.
* Customised font-weight styling of title vs meta info.

This is the use case you mention and it is also what I mean when I say we need to be able to **customise the rendering of suggestions** that `URLInput` provides. 

The current implementation is one route and it looks like you've now provided another option which is great.

> My feeling is that as a first step It'd be far more preferable to provide a simple api:

Thanks for providing a concrete suggestion. This is where _I_ was missing the context.

I don't know which option is best. Your option does couple `URLInput` to the `LinkControl`, but it would be experimental so less problematic if we needed to reverse out of it in the future. Another upside of your approach is that we don't expose an API on `URLInput` which could lead to inaccessible implementations.

I feel like we need someone to make a call on this - @youknowriad can you offer any advice? I'm happy either way.

Thanks again for persisting with this.

Reply:
It looks like this conundrum was resolved in https://github.com/WordPress/gutenberg/issues/17557#issuecomment-544894801? The idea is to merge this as experimental (we should update the nomenclature to make that clear) and evaluate consolidating after trying it out separately.

I took the liberty to update the title of this PR to help clarify the intent.